File: ThirdPartyAdapters/vungle/vungle/src/main/java/com/google/ads/mediation/vungle/VungleMediationAdapter.java
package com.google.ads.mediation.vungle;

import android.content.Context;
import android.os.Bundle;
import android.text.TextUtils;
import android.util.Log;

import androidx.annotation.IntDef;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.google.ads.mediation.vungle.VungleInitializer.VungleInitializationListener;
import com.google.ads.mediation.vungle.rtb.VungleRtbInterstitialAd;
import com.google.ads.mediation.vungle.rtb.VungleRtbRewardedAd;
import com.google.android.gms.ads.AdError;
import com.google.android.gms.ads.mediation.InitializationCompleteCallback;
import com.google.android.gms.ads.mediation.MediationAdLoadCallback;
import com.google.android.gms.ads.mediation.MediationConfiguration;
import com.google.android.gms.ads.mediation.MediationNativeAdCallback;
import com.google.android.gms.ads.mediation.MediationNativeAdConfiguration;
import com.google.android.gms.ads.mediation.MediationInterstitialAd;
import com.google.android.gms.ads.mediation.MediationInterstitialAdCallback;
import com.google.android.gms.ads.mediation.MediationInterstitialAdConfiguration;
import com.google.android.gms.ads.mediation.MediationRewardedAd;
import com.google.android.gms.ads.mediation.MediationRewardedAdCallback;
import com.google.android.gms.ads.mediation.MediationRewardedAdConfiguration;
import com.google.android.gms.ads.mediation.UnifiedNativeAdMapper;
import com.google.android.gms.ads.mediation.VersionInfo;
import com.google.android.gms.ads.mediation.rtb.RtbAdapter;
import com.google.android.gms.ads.mediation.rtb.RtbSignalData;
import com.google.android.gms.ads.mediation.rtb.SignalCallbacks;
import com.google.android.gms.ads.rewarded.RewardItem;
import com.vungle.ads.AdConfig;
import com.vungle.ads.BannerAd;
import com.vungle.ads.BaseAd;
import com.vungle.ads.BaseAdListener;
import com.vungle.ads.InitializationListener;
import com.vungle.ads.RewardedAd;
import com.vungle.ads.VungleAds;
import com.vungle.ads.VungleException;
import com.vungle.ads.VungleSettings;
import com.vungle.mediation.BuildConfig;
import com.vungle.mediation.VungleExtrasBuilder;
import com.vungle.mediation.VungleManager;
import com.vungle.mediation.VungleNativeAdapter;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.ref.WeakReference;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;

/**
 * Mediation network adapter for Vungle.
 */
public class VungleMediationAdapter extends RtbAdapter {

    public static final String TAG = VungleMediationAdapter.class.getSimpleName();
    public static final String KEY_APP_ID = "appid";

    private VungleRtbInterstitialAd rtbInterstitialAd;
    private VungleRtbRewardedAd rtbRewardedAd;
    private VungleRtbRewardedAd rtbRewardedInterstitialAd;

    private AdConfig mAdConfig;
    private String mUserID;
    private String mPlacement;

    private static final HashMap<String, WeakReference<VungleMediationAdapter>> mPlacementsInUse =
            new HashMap<>();

    private MediationAdLoadCallback<MediationRewardedAd, MediationRewardedAdCallback>
            mMediationAdLoadCallback;
    private MediationRewardedAdCallback mMediationRewardedAdCallback;

    /**
     * Vungle adapter error domain.
     */
    public static final String ERROR_DOMAIN = "com.google.ads.mediation.vungle";

    @Retention(RetentionPolicy.SOURCE)
    @IntDef(
            value = {
                    ERROR_INVALID_SERVER_PARAMETERS,
                    ERROR_BANNER_SIZE_MISMATCH,
                    ERROR_REQUIRES_ACTIVITY_CONTEXT,
                    ERROR_AD_ALREADY_LOADED,
                    ERROR_VUNGLE_BANNER_NULL,
                    ERROR_INITIALIZATION_FAILURE,
                    ERROR_CANNOT_PLAY_AD,
            })

    public @interface AdapterError {

    }

    /**
     * Server parameters, such as app ID or placement ID, are invalid.
     */
    public static final int ERROR_INVALID_SERVER_PARAMETERS = 101;

    /**
     * The requested ad size does not match a Vungle supported banner size.
     */
    public static final int ERROR_BANNER_SIZE_MISMATCH = 102;

    /**
     * Vungle requires an {@link android.app.Activity} context to request ads.
     */
    public static final int ERROR_REQUIRES_ACTIVITY_CONTEXT = 103;

    /**
     * Vungle SDK cannot load multiple ads for the same placement ID.
     */
    public static final int ERROR_AD_ALREADY_LOADED = 104;

    /**
     * Vungle SDK failed to initialize.
     */
    public static final int ERROR_INITIALIZATION_FAILURE = 105;

    /**
     * Vungle SDK returned a successful load callback, but Banners.getBanner() or Vungle.getNativeAd()
     * returned null.
     */
    public static final int ERROR_VUNGLE_BANNER_NULL = 106;

    /**
     * Vungle SDK is not ready to play the ad.
     */
    public static final int ERROR_CANNOT_PLAY_AD = 107;

    @NonNull
    @Override
    public VersionInfo getVersionInfo() {
        String versionString = BuildConfig.ADAPTER_VERSION;
        String[] splits = versionString.split("\\.");

        if (splits.length >= 4) {
            int major = Integer.parseInt(splits[0]);
            int minor = Integer.parseInt(splits[1]);
            int micro = Integer.parseInt(splits[2]) * 100 + Integer.parseInt(splits[3]);
            return new VersionInfo(major, minor, micro);
        }

        String logMessage =
                String.format(
                        "Unexpected adapter version format: %s. Returning 0.0.0 for adapter version.",
                        versionString);
        Log.w(TAG, logMessage);
        return new VersionInfo(0, 0, 0);
    }

    @NonNull
    @Override
    public VersionInfo getSDKVersionInfo() {
        String versionString = com.vungle.ads.BuildConfig.VERSION_NAME;
        String[] splits = versionString.split("\\.");

        if (splits.length >= 3) {
            int major = Integer.parseInt(splits[0]);
            int minor = Integer.parseInt(splits[1]);
            int micro = Integer.parseInt(splits[2]);
            return new VersionInfo(major, minor, micro);
        }

        String logMessage =
                String.format(
                        "Unexpected SDK version format: %s. Returning 0.0.0 for SDK version.", versionString);
        Log.w(TAG, logMessage);
        return new VersionInfo(0, 0, 0);
    }

    @Override
    public void collectSignals(@NonNull RtbSignalData rtbSignalData,
                               @NonNull SignalCallbacks signalCallbacks) {
//        String token = Vungle.getAvailableBidTokens(rtbSignalData.getContext());
//        Log.d(TAG, "token=" + token);
//        signalCallbacks.onSuccess(token);
    }

    @Override
    public void initialize(@NonNull Context context,
                           @NonNull final InitializationCompleteCallback initializationCompleteCallback,
                           @NonNull List<MediationConfiguration> mediationConfigurations) {

        if (VungleAds.isInitialized()) {
            initializationCompleteCallback.onInitializationSucceeded();
            return;
        }
        HashSet<String> appIDs = new HashSet<>();
        for (MediationConfiguration configuration : mediationConfigurations) {
            Bundle serverParameters = configuration.getServerParameters();
            String appIDFromServer = serverParameters.getString(KEY_APP_ID);

            if (!TextUtils.isEmpty(appIDFromServer)) {
                appIDs.add(appIDFromServer);
            }
        }

        int count = appIDs.size();
        if (count <= 0) {
            if (initializationCompleteCallback != null) {
                AdError error = new AdError(ERROR_INVALID_SERVER_PARAMETERS, "Missing or Invalid App ID.",
                        ERROR_DOMAIN);
                Log.w(TAG, error.getMessage());
                initializationCompleteCallback.onInitializationFailed(error.getMessage());
            }
            return;
        }

        String appID = appIDs.iterator().next();
        if (count > 1) {
            String logMessage =
                    String.format(
                            "Multiple '%s' entries found: %s. Using '%s' to initialize the Vungle SDK.",
                            KEY_APP_ID, appIDs, appID);
            Log.w(TAG, logMessage);
        }

        VungleAds.init(
                context,
                appID,
                new InitializationListener() {

                    @Override
                    public void onSuccess() {
                        initializationCompleteCallback.onInitializationSucceeded();
                    }

                    @Override
                    public void onError(@NonNull VungleException e) {
                        Log.w(TAG, e.getMessage());
                        initializationCompleteCallback.onInitializationFailed(e.getMessage());
                    }
                },
                new VungleSettings()
        );
    }

    @Override
    public void loadRewardedAd(
            @NonNull MediationRewardedAdConfiguration mediationRewardedAdConfiguration,
            @NonNull MediationAdLoadCallback<MediationRewardedAd, MediationRewardedAdCallback>
                    mediationAdLoadCallback) {
        mMediationAdLoadCallback = mediationAdLoadCallback;

        Bundle mediationExtras = mediationRewardedAdConfiguration.getMediationExtras();
        Bundle serverParameters = mediationRewardedAdConfiguration.getServerParameters();

        mUserID = mediationExtras.getString(VungleExtrasBuilder.EXTRA_USER_ID);

        mPlacement = VungleManager.getInstance().findPlacement(mediationExtras, serverParameters);
        if (TextUtils.isEmpty(mPlacement)) {
            AdError error = new AdError(ERROR_INVALID_SERVER_PARAMETERS,
                    "Failed to load ad from Vungle. Missing or invalid Placement ID.", ERROR_DOMAIN);
            Log.w(TAG, error.getMessage());
            mediationAdLoadCallback.onFailure(error);
            return;
        }

        if (mPlacementsInUse.containsKey(mPlacement)
                && mPlacementsInUse.get(mPlacement).get() != null) {
            AdError error = new AdError(ERROR_AD_ALREADY_LOADED,
                    "Only a maximum of one ad can be loaded per placement.", ERROR_DOMAIN);
            Log.w(TAG, error.getMessage());
            mediationAdLoadCallback.onFailure(error);
            return;
        }

        String appID = serverParameters.getString(KEY_APP_ID);
        if (TextUtils.isEmpty(appID)) {
            AdError error = new AdError(ERROR_INVALID_SERVER_PARAMETERS,
                    "Failed to load ad from Vungle. Missing or Invalid App ID.", ERROR_DOMAIN);
            Log.w(TAG, error.getMessage());
            mediationAdLoadCallback.onFailure(error);
            return;
        }

        // Unmute full-screen ads by default.
        mAdConfig = VungleExtrasBuilder.adConfigWithNetworkExtras(mediationExtras, false);

        VungleRtbRewardedAd ad = new VungleRtbRewardedAd(
                mediationRewardedAdConfiguration,
                mediationAdLoadCallback
        );
        ad.render();
    }

    /**
     * This class is used to map Vungle rewarded video ad rewards to Google Mobile Ads SDK rewards.
     */
    public static class VungleReward implements RewardItem {

        private final String mType;
        private final int mAmount;

        public VungleReward(String type, int amount) {
            mType = type;
            mAmount = amount;
        }

        @Override
        public int getAmount() {
            return mAmount;
        }

        @NonNull
        @Override
        public String getType() {
            return mType;
        }
    }

    @Override
    public void loadNativeAd(@NonNull MediationNativeAdConfiguration mediationNativeAdConfiguration,
                             @NonNull MediationAdLoadCallback<UnifiedNativeAdMapper, MediationNativeAdCallback> callback) {
        Log.d(TAG, "loadNativeAd()...");
        VungleInitializer.getInstance()
                .updateCoppaStatus(mediationNativeAdConfiguration.taggedForChildDirectedTreatment());
        VungleNativeAdapter nativeAdapter = new VungleNativeAdapter(mediationNativeAdConfiguration,
                callback);
        nativeAdapter.render();
    }

    public void loadRtbRewardedAd(@NonNull MediationRewardedAdConfiguration mediationRewardedAdConfiguration,
                                  @NonNull MediationAdLoadCallback<MediationRewardedAd, MediationRewardedAdCallback> mediationAdLoadCallback) {
        Log.d(TAG, "loadRtbRewardedAd()...");
        VungleInitializer.getInstance()
                .updateCoppaStatus(mediationRewardedAdConfiguration.taggedForChildDirectedTreatment());
        rtbRewardedAd = new VungleRtbRewardedAd(
                mediationRewardedAdConfiguration, mediationAdLoadCallback);
        rtbRewardedAd.render();
    }

    @Override
    public void loadRtbInterstitialAd(
            @NonNull MediationInterstitialAdConfiguration mediationInterstitialAdConfiguration,
            @NonNull MediationAdLoadCallback<MediationInterstitialAd, MediationInterstitialAdCallback> mediationAdLoadCallback) {
        Log.d(TAG, "loadRtbInterstitialAd()...");
        VungleInitializer.getInstance()
                .updateCoppaStatus(mediationInterstitialAdConfiguration.taggedForChildDirectedTreatment());
        rtbInterstitialAd = new VungleRtbInterstitialAd(
                mediationInterstitialAdConfiguration, mediationAdLoadCallback);
        rtbInterstitialAd.render();
    }

    @Override
    public void loadRtbNativeAd(@NonNull MediationNativeAdConfiguration adConfiguration,
                                @NonNull MediationAdLoadCallback<UnifiedNativeAdMapper, MediationNativeAdCallback> callback) {
        Log.d(TAG, "loadRtbNativeAd()...");
        VungleInitializer.getInstance()
                .updateCoppaStatus(adConfiguration.taggedForChildDirectedTreatment());
        VungleNativeAdapter nativeAdapter = new VungleNativeAdapter(adConfiguration, callback);
        nativeAdapter.render();
    }

    @Override
    public void loadRtbRewardedInterstitialAd(
            @NonNull MediationRewardedAdConfiguration adConfiguration,
            @NonNull MediationAdLoadCallback<MediationRewardedAd, MediationRewardedAdCallback> callback) {
        Log.d(TAG, "loadRtbRewardedInterstitialAd()...");
        VungleInitializer.getInstance()
                .updateCoppaStatus(adConfiguration.taggedForChildDirectedTreatment());
        rtbRewardedInterstitialAd = new VungleRtbRewardedAd(adConfiguration, callback);
        rtbRewardedInterstitialAd.render();
    }

    @Override
    public void loadRewardedInterstitialAd(
            @NonNull MediationRewardedAdConfiguration mediationRewardedAdConfiguration,
            @NonNull MediationAdLoadCallback<MediationRewardedAd, MediationRewardedAdCallback> callback) {
        Log.d(TAG, "loadRewardedInterstitialAd()...");
        loadRewardedAd(mediationRewardedAdConfiguration, callback);
    }
}

File: ThirdPartyAdapters/vungle/vungle/src/main/java/com/vungle/mediation/VungleMediationAdapter.kt
package com.vungle.mediation

import android.content.Context
import android.text.TextUtils
import android.util.Log
import androidx.annotation.IntDef
import com.google.ads.mediation.vungle.VungleInitializer
import com.google.ads.mediation.vungle.rtb.VungleRtbInterstitialAd
import com.google.android.gms.ads.AdError
import com.google.android.gms.ads.mediation.*
import com.google.android.gms.ads.mediation.rtb.RtbAdapter
import com.google.android.gms.ads.mediation.rtb.RtbSignalData
import com.google.android.gms.ads.mediation.rtb.SignalCallbacks
import com.google.android.gms.ads.rewarded.RewardItem
import com.vungle.ads.*
import java.lang.annotation.Retention
import java.lang.annotation.RetentionPolicy
import java.lang.ref.WeakReference

/**
 * Mediation network adapter for Vungle.
 */
open class VungleMediationAdapter : RtbAdapter(), MediationRewardedAd {
    private var rtbInterstitialAd: VungleRtbInterstitialAd? = null
//    private var rtbRewardedAd: VungleRtbRewardedAd? = null
//    private var rtbRewardedInterstitialAd: VungleRtbRewardedAd? = null
    private var mAdConfig: AdConfig? = null
    private var mUserID: String? = null
    private var mPlacement: String? = null
    private var mMediationAdLoadCallback: MediationAdLoadCallback<MediationRewardedAd, MediationRewardedAdCallback>? =
        null
    private var mMediationRewardedAdCallback: MediationRewardedAdCallback? = null

    @Retention(RetentionPolicy.SOURCE)
    @IntDef(value = [ERROR_INVALID_SERVER_PARAMETERS, ERROR_BANNER_SIZE_MISMATCH, ERROR_REQUIRES_ACTIVITY_CONTEXT, ERROR_AD_ALREADY_LOADED, ERROR_VUNGLE_BANNER_NULL, ERROR_INITIALIZATION_FAILURE, ERROR_CANNOT_PLAY_AD])
    annotation class AdapterError

    override fun getVersionInfo(): VersionInfo {
        val versionString = BuildConfig.ADAPTER_VERSION
        val splits = versionString.split("\\.").toTypedArray()
        if (splits.size >= 4) {
            val major = splits[0].toInt()
            val minor = splits[1].toInt()
            val micro = splits[2].toInt() * 100 + splits[3].toInt()
            return VersionInfo(major, minor, micro)
        }
        val logMessage = String.format(
            "Unexpected adapter version format: %s. Returning 0.0.0 for adapter version.",
            versionString
        )
        Log.w(TAG, logMessage)
        return VersionInfo(0, 0, 0)
    }

    override fun getSDKVersionInfo(): VersionInfo {
        val versionString: String = com.vungle.ads.BuildConfig.VERSION_NAME
        val splits = versionString.split("\\.").toTypedArray()
        if (splits.size >= 3) {
            val major = splits[0].toInt()
            val minor = splits[1].toInt()
            val micro = splits[2].toInt()
            return VersionInfo(major, minor, micro)
        }
        val logMessage = String.format(
            "Unexpected SDK version format: %s. Returning 0.0.0 for SDK version.", versionString
        )
        Log.w(TAG, logMessage)
        return VersionInfo(0, 0, 0)
    }

    override fun collectSignals(
        rtbSignalData: RtbSignalData,
        signalCallbacks: SignalCallbacks
    ) {
        val token = VungleAds.getBidTokens()
        Log.d(TAG, "token=$token")
        signalCallbacks.onSuccess(token)
    }

    override fun initialize(
        context: Context,
        initializationCompleteCallback: InitializationCompleteCallback,
        mediationConfigurations: List<MediationConfiguration>
    ) {
        if (VungleAds.isInitialized()) {
            initializationCompleteCallback.onInitializationSucceeded()
            return
        }
        val appIDs = HashSet<String?>()
        for (configuration in mediationConfigurations) {
            val serverParameters = configuration.serverParameters
            val appIDFromServer = serverParameters.getString(KEY_APP_ID)
            if (!TextUtils.isEmpty(appIDFromServer)) {
                appIDs.add(appIDFromServer)
            }
        }
        val count = appIDs.size
        if (count <= 0) {
            val error = AdError(
                ERROR_INVALID_SERVER_PARAMETERS, "Missing or Invalid App ID.",
                ERROR_DOMAIN
            )
            Log.w(TAG, error.message)
            initializationCompleteCallback.onInitializationFailed(error.message)
            return
        }
        val appID = appIDs.iterator().next()
        if (count > 1) {
            val logMessage = String.format(
                "Multiple '%s' entries found: %s. Using '%s' to initialize the Vungle SDK.",
                KEY_APP_ID, appIDs, appID
            )
            Log.w(TAG, logMessage)
        }
        if (appID == null) {
            return
            // TODO: can this be null
        }
//        VungleAds.init(context.applicationContext, appID, object : InitializationListener {
//            override fun onError(vungleException: VungleException) {
//                Log.w(TAG, vungleException.message!!)
//                initializationCompleteCallback.onInitializationFailed(error.message)
//            }
//
//            override fun onSuccess() {
//                initializationCompleteCallback.onInitializationSucceeded()
//            }
//
//        }, VungleSettings())
    }

    override fun loadRewardedAd(
        mediationRewardedAdConfiguration: MediationRewardedAdConfiguration,
        mediationAdLoadCallback: MediationAdLoadCallback<MediationRewardedAd, MediationRewardedAdCallback>
    ) {
//        mMediationAdLoadCallback = mediationAdLoadCallback
//        val mediationExtras = mediationRewardedAdConfiguration.mediationExtras
//        val serverParameters = mediationRewardedAdConfiguration.serverParameters
//        if (mediationExtras != null) {
//            mUserID = mediationExtras.getString(VungleExtrasBuilder.EXTRA_USER_ID)
//        }
//        mPlacement = VungleManager.getInstance().findPlacement(mediationExtras, serverParameters)
//        if (TextUtils.isEmpty(mPlacement)) {
//            val error = AdError(
//                ERROR_INVALID_SERVER_PARAMETERS,
//                "Failed to load ad from Vungle. Missing or invalid Placement ID.", ERROR_DOMAIN
//            )
//            Log.w(TAG, error.message)
//            mediationAdLoadCallback.onFailure(error)
//            return
//        }
//        if (mPlacementsInUse.containsKey(mPlacement)
//            && mPlacementsInUse[mPlacement]!!.get() != null
//        ) {
//            val error = AdError(
//                ERROR_AD_ALREADY_LOADED,
//                "Only a maximum of one ad can be loaded per placement.", ERROR_DOMAIN
//            )
//            Log.w(TAG, error.message)
//            mediationAdLoadCallback.onFailure(error)
//            return
//        }
//        val appID = serverParameters.getString(KEY_APP_ID)
//        if (TextUtils.isEmpty(appID)) {
//            val error = AdError(
//                ERROR_INVALID_SERVER_PARAMETERS,
//                "Failed to load ad from Vungle. Missing or Invalid App ID.", ERROR_DOMAIN
//            )
//            Log.w(TAG, error.message)
//            mediationAdLoadCallback.onFailure(error)
//            return
//        }
//
//        // Unmute full-screen ads by default.
//        mAdConfig = VungleExtrasBuilder.adConfigWithNetworkExtras(mediationExtras, false)
//        VungleInitializer.instance
//            .updateCoppaStatus(mediationRewardedAdConfiguration.taggedForChildDirectedTreatment())
//        VungleInitializer.instance
//            .initialize(
//                appID,
//                mediationRewardedAdConfiguration.context,
//                object : VungleInitializationListener() {
//                    fun onInitializeSuccess() {
//                        Vungle.setIncentivizedFields(mUserID, null, null, null, null)
//                        mPlacementsInUse[mPlacement] = WeakReference(this@VungleMediationAdapter)
//                        if (Vungle.canPlayAd(mPlacement)) {
//                            mMediationRewardedAdCallback =
//                                mMediationAdLoadCallback!!.onSuccess(this@VungleMediationAdapter)
//                            return
//                        }
//                        Vungle.loadAd(mPlacement, mAdConfig, this@VungleMediationAdapter)
//                    }
//
//                    fun onInitializeError(error: AdError) {
//                        Log.w(TAG, error.message)
//                        mMediationAdLoadCallback!!.onFailure(error)
//                        mPlacementsInUse.remove(mPlacement)
//                    }
//                })
    }

    override fun showAd(context: Context) {
//        Vungle.playAd(mPlacement, mAdConfig, this@VungleMediationAdapter)
    }

    /**
     * [LoadAdCallback] implementation from Vungle.
     */
    fun onAdLoad(placementId: String?) {
        if (mMediationAdLoadCallback != null) {
            mMediationRewardedAdCallback =
                mMediationAdLoadCallback!!.onSuccess(this@VungleMediationAdapter)
        }
        mPlacementsInUse[mPlacement] =
            WeakReference(this@VungleMediationAdapter)
    }

    fun creativeId(creativeId: String?) {
        // no-op
    }

    /**
     * [PlayAdCallback] implementation from Vungle
     */
    fun onAdStart(placementId: String?) {
        if (mMediationRewardedAdCallback != null) {
            mMediationRewardedAdCallback!!.onAdOpened()
        }
    }

    @Deprecated("")
    fun onAdEnd(
        placementId: String?, wasSuccessfulView: Boolean,
        wasCallToActionClicked: Boolean
    ) {
    }

    fun onAdEnd(placementId: String?) {
        if (mMediationRewardedAdCallback != null) {
            mMediationRewardedAdCallback!!.onAdClosed()
        }
        mPlacementsInUse.remove(placementId)
    }

    fun onAdClick(placementId: String?) {
        if (mMediationRewardedAdCallback != null) {
            mMediationRewardedAdCallback!!.reportAdClicked()
        }
    }

    fun onAdRewarded(placementId: String?) {
        if (mMediationRewardedAdCallback != null) {
            mMediationRewardedAdCallback!!.onVideoComplete()
            mMediationRewardedAdCallback!!.onUserEarnedReward(VungleReward("vungle", 1))
        }
    }

    fun onAdLeftApplication(placementId: String?) {
        // no op
    }

    // Vungle's LoadAdCallback and PlayAdCallback shares the same onError() call; when an
    // ad request to Vungle fails, and when an ad fails to play.
    fun onError(placementId: String?, throwable: VungleException) {
        val error = getAdError(throwable)
        Log.w(TAG, error.message)
        if (mMediationRewardedAdCallback != null) {
            mMediationRewardedAdCallback!!.onAdFailedToShow(error)
        } else if (mMediationAdLoadCallback != null) {
            mMediationAdLoadCallback!!.onFailure(error)
        }
        mPlacementsInUse.remove(placementId)
    }

    fun onAdViewed(placementId: String?) {
        mMediationRewardedAdCallback!!.onVideoStart()
        mMediationRewardedAdCallback!!.reportAdImpression()
    }

    /**
     * This class is used to map Vungle rewarded video ad rewards to Google Mobile Ads SDK rewards.
     */
    class VungleReward(private val mType: String, private val mAmount: Int) : RewardItem {
        override fun getAmount(): Int {
            return mAmount
        }

        override fun getType(): String {
            return mType
        }
    }

    override fun loadNativeAd(
        mediationNativeAdConfiguration: MediationNativeAdConfiguration,
        callback: MediationAdLoadCallback<UnifiedNativeAdMapper, MediationNativeAdCallback>
    ) {
        Log.d(TAG, "loadNativeAd()...")
        VungleInitializer.updateCoppaStatus(mediationNativeAdConfiguration.taggedForChildDirectedTreatment())
//        val nativeAdapter = VungleNativeAdapter(
//            mediationNativeAdConfiguration,
//            callback
//        )
//        nativeAdapter.render()
    }

    override fun loadRtbRewardedAd(
        mediationRewardedAdConfiguration: MediationRewardedAdConfiguration,
        mediationAdLoadCallback: MediationAdLoadCallback<MediationRewardedAd, MediationRewardedAdCallback>
    ) {
        // TODO: replace:
//        Log.d(TAG, "loadRtbRewardedAd()...")
//        VungleInitializer.updateCoppaStatus(mediationRewardedAdConfiguration.taggedForChildDirectedTreatment())
//        rtbRewardedAd = VungleRtbRewardedAd(
//            mediationRewardedAdConfiguration, mediationAdLoadCallback
//        )
//        rtbRewardedAd!!.render()
    }

    override fun loadRtbInterstitialAd(
        mediationInterstitialAdConfiguration: MediationInterstitialAdConfiguration,
        mediationAdLoadCallback: MediationAdLoadCallback<MediationInterstitialAd, MediationInterstitialAdCallback>
    ) {
        Log.d(TAG, "loadRtbInterstitialAd()...")
        VungleInitializer
            .updateCoppaStatus(mediationInterstitialAdConfiguration.taggedForChildDirectedTreatment())
        rtbInterstitialAd = VungleRtbInterstitialAd(
            mediationInterstitialAdConfiguration, mediationAdLoadCallback
        )
        rtbInterstitialAd!!.render()
    }

    override fun loadRtbNativeAd(
        adConfiguration: MediationNativeAdConfiguration,
        callback: MediationAdLoadCallback<UnifiedNativeAdMapper, MediationNativeAdCallback>
    ) {
        Log.d(TAG, "loadRtbNativeAd()...")
        VungleInitializer.updateCoppaStatus(adConfiguration.taggedForChildDirectedTreatment())
//        val nativeAdapter = VungleNativeAdapter(adConfiguration, callback)
//        nativeAdapter.render()
    }

    override fun loadRtbRewardedInterstitialAd(
        adConfiguration: MediationRewardedAdConfiguration,
        callback: MediationAdLoadCallback<MediationRewardedAd, MediationRewardedAdCallback>
    ) {
        Log.d(TAG, "loadRtbRewardedInterstitialAd()...")
        VungleInitializer.updateCoppaStatus(adConfiguration.taggedForChildDirectedTreatment())
//        rtbRewardedInterstitialAd = VungleRtbRewardedAd(adConfiguration, callback)
//        rtbRewardedInterstitialAd.render()
    }

    override fun loadRewardedInterstitialAd(
        mediationRewardedAdConfiguration: MediationRewardedAdConfiguration,
        callback: MediationAdLoadCallback<MediationRewardedAd, MediationRewardedAdCallback>
    ) {
        Log.d(TAG, "loadRewardedInterstitialAd()...")
        loadRewardedAd(mediationRewardedAdConfiguration, callback)
    }

    companion object {
        @JvmField
        val TAG = VungleMediationAdapter::class.java.simpleName
        const val KEY_APP_ID = "appid"
        private val mPlacementsInUse = HashMap<String?, WeakReference<VungleMediationAdapter?>>()

        /**
         * Vungle adapter error domain.
         */
        const val ERROR_DOMAIN = "com.google.ads.mediation.vungle"

        /**
         * Vungle SDK error domain.
         */
        const val VUNGLE_SDK_ERROR_DOMAIN = "com.vungle.warren"

        /**
         * Server parameters, such as app ID or placement ID, are invalid.
         */
        const val ERROR_INVALID_SERVER_PARAMETERS = 101

        /**
         * The requested ad size does not match a Vungle supported banner size.
         */
        const val ERROR_BANNER_SIZE_MISMATCH = 102

        /**
         * Vungle requires an [android.app.Activity] context to request ads.
         */
        const val ERROR_REQUIRES_ACTIVITY_CONTEXT = 103

        /**
         * Vungle SDK cannot load multiple ads for the same placement ID.
         */
        const val ERROR_AD_ALREADY_LOADED = 104

        /**
         * Vungle SDK failed to initialize.
         */
        const val ERROR_INITIALIZATION_FAILURE = 105

        /**
         * Vungle SDK returned a successful load callback, but Banners.getBanner() or Vungle.getNativeAd()
         * returned null.
         */
        const val ERROR_VUNGLE_BANNER_NULL = 106

        /**
         * Vungle SDK is not ready to play the ad.
         */
        const val ERROR_CANNOT_PLAY_AD = 107

        /**
         * Convert the given Vungle exception into the appropriate custom error code.
         */
        @JvmStatic
        fun getAdError(exception: VungleException): AdError {
            return AdError(
                exception.exceptionCode, exception.localizedMessage,
                VUNGLE_SDK_ERROR_DOMAIN
            )
        }
    }
}